src/server/lib/http-request.js

```diff
@@ -0,0 +1,123 @@
+import originalFetch from "node-fetch";
+import AbortController from "node-abort-controller";
+import { log } from "../../lib";
+import { v4 as uuid } from "uuid";
+
+const requestWithRetry = async (url, props) => {
+  const originalProps = props || {};
+  const remainingProps = {
+    ...(originalProps || {})
+  };
+  delete remainingProps.validateStatus;
+  delete remainingProps.retries;
+  delete remainingProps.timeout;
+
+  const requestId = uuid();
+  const propsRetries = originalProps.retries;
+  const retries =
+    propsRetries !== null && propsRetries !== undefined ? propsRetries : 2;
+  const timeout = originalProps.timeout || 2000;
+
+  const retryDelay = () => {
+    const baseDelay = 50;
+    const randomDelay = Math.floor(Math.random() * (baseDelay / 2));
+    return baseDelay + randomDelay;
+  };
+
+  const validateStatus = status => {
+    const validateProp = originalProps.validateStatus;
+    let acceptableStatuses = [200];
+    if (validateProp) {
+      if (typeof validateProp === "function") {
+        return validateProp(status);
+      } else if (Array.isArray(validateProp)) {
+        acceptableStatuses = validateProp;
+      } else {
+        acceptableStatuses = [validateProp];
+      }
+    }
+
+    return acceptableStatuses.includes(status);
+  };
+
+  const logRequest = () => {
+    return JSON.stringify({ url, props });
+  };
+
+  const RetryReturnError = {
+    RETURN: 1,
+    RETRY: 2,
+    ERROR: 3
+  };
+  const shouldRetryReturnOrError = (attempt, error, resp) => {
+    if (validateStatus(resp && resp.status)) {
+      return RetryReturnError.RETURN;
+    }
+
+    if (
+      attempt < retries &&
+      (error || (resp && resp.status >= 500 && resp.status <= 599))
+    ) {
+      let message;
+      if (error) {
+        let tweakedError = error;
+        if (error.toString && error.toString().match(/.*AbortError.*/g)) {
+          tweakedError = `timeout after ${timeout}ms`;
+        }
+        message = `error -- ${tweakedError}`;
+      } else {
+        message = `status code ${resp.status}`;
+      }
+
+      log.warn(
+        `Retrying request id ${requestId}. Reason: ${message}. Details: ${logRequest()}. Retry ${attempt +
+          1} of ${retries}`
+      );
+      return RetryReturnError.RETRY;
+    }
+
+    return RetryReturnError.ERROR;
+  };
+
+  for (let attempt = 0; attempt < retries + 1; attempt++) {
+    let error;
+    let response;
+
+    const controller = new AbortController();
+    const controllerTimeout = setInterval(() => {
```

Comment:
Shouldn't this be `setTimeout`?  if not, then you don't need to 'clear' it below, and this is inside the loop, so it's just once, right?  If not, then please document this extensively.

Reply:
Yeah, good catch.

Reply:
`clearTimeout` is necessary for when the request completes before the timer timeout fires. It's inside the loop because a request can fail, and be retried, and the retry could time out.

Reply:
SHA 04fd3889d999dcaf618a21b749fa95b772df50f6